### ISSUE TITLE
Transaction get, get_range return types

### ISSUE THREAD

**Issue by Speedy37** [bluejekyll/foundationdb-rs#93]
Hi,

I was wondering about the Transaction api:

Why is `Transaction.get` future not directly returning `Option<Vec<u8>>` ?

I propose:

```rust
fn get(&self, key: &[u8], snapshot: bool) -> Future<Option<&[u8]>>;
fn get_range(&self, opt: RangeOption) -> Stream<KeyValue>;
fn get_key(&self, selector: KeySelector, snapshot: bool) -> Future<&[u8]>;
```

What do you think ?

**Comment by yjh0502** [comment]
There are several considerations on current API.

 - You might want to get additional information about the query after fetching the value. For example, https://docs.rs/foundationdb/0.1.0/foundationdb/transaction/struct.Transaction.html#method.committed_version

 - Current API does not copy a value from fdb api, and a caller can copy it if they need to. If we return Vec, the value will be always copied.
 - It's common to call multiple read commands, or call write command after read. With current API, caller can get a `Transaction` which can be used to continue operation with same transaction. If API returns value only, caller need to clone the transactions and carry it around closures.

**Comment by Speedy37** [comment]
Point 1 & 3 are about retaining the transaction object: there is one indirection just for that case, while the transaction object is easy to clone. I don't see this as an advantage.

Point 2, I meant to return `&[u8]` (probably require some lifetime declaration with it: as long as the Future is alive, &[u8] is alive).

**Comment by Speedy37** [comment]
For `get_range`, I can imagine a need to be able to manually access "bunchs" of key/value but I think the most useful case is just: iterate and at any point in the iteration being able to "generate" a "next" option for future calls to `get_range` (for pagination).

**Comment by bluejekyll** [comment]
So basically you want to drop the `Transaction` object as part of the `GetResult` future result from the `get`s, right? And then the user would need to clone `Transaction` if they plan to use this in any other lambdas or future combinators, right?

Another option would be to make `GetResult` (and the others) simply implement `Deref<Result<Option<&[u8]>>` and then it would appear to have the interface you're asking for, right? That would be the most minimal change and leave the chaining in place that @yjh0502 likes.

**Comment by Speedy37** [comment]
Right. It's a tradeoff I would take. I'll make a real life example with both cases.

Actually I prefer `value()` to deref, it's just to remove the "double Result" check (by the future and inside then).

**Comment by bluejekyll** [comment]
yeah, the Result is awkward there... I wonder if we could pull that up into the future and return it as it's Error type.

**Comment by yjh0502** [comment]
Result type is came from `fdb::fdb_future_get_value`, which returns an error if a future does not have a value. I think we can remove that if we check future type on FdbFuture, and we are already doing that on some place, like https://github.com/bluejekyll/foundationdb-rs/blob/master/foundationdb/src/transaction.rs

**Comment by bluejekyll** [comment]
I just merged #94

**Comment by bluejekyll** [comment]
This was fixed a while back, closing.
